File: cmd/codegen/generator/go/templates/format.go
package templates

import (
	"strings"

	"github.com/dagger/dagger/cmd/codegen/generator"
	"github.com/dagger/dagger/cmd/codegen/introspection"
)

// FormatTypeFunc is an implementation of generator.FormatTypeFuncs interface
// to format GraphQL type into Golang.
type FormatTypeFunc struct {
	scope string
}

func (f *FormatTypeFunc) WithScope(scope string) generator.FormatTypeFuncs {
	if scope != "" {
		scope += "."
	}
	clone := *f
	clone.scope = scope
	return &clone
}

func (f *FormatTypeFunc) FormatKindList(representation string) string {
	representation = "[]" + representation
	return representation
}

func (f *FormatTypeFunc) FormatKindScalarString(representation string) string {
	representation += "string"
	return representation
}

func (f *FormatTypeFunc) FormatKindScalarInt(representation string) string {
	representation += "int"
	return representation
}

func (f *FormatTypeFunc) FormatKindScalarFloat(representation string) string {
	representation += "float64"
	return representation
}

func (f *FormatTypeFunc) FormatKindScalarBoolean(representation string) string {
	representation += "bool"
	return representation
}

func (f *FormatTypeFunc) FormatKindScalarDefault(representation string, refName string, input bool) string {
	if obj, ok := strings.CutSuffix(refName, "ID"); input && ok && obj != "" {
		// Check if the object type is an interface. If so, the Go type
		// is a Go interface (not a struct), so we don't add a pointer prefix.
		if schema := generator.GetSchema(); schema != nil {
			if schemaType := schema.Types.Get(obj); schemaType != nil && schemaType.Kind == introspection.TypeKindInterface {
				representation += f.scope + formatName(obj)
				return representation
			}
		}
		representation += "*" + f.scope + formatName(obj)
	} else {
		representation += f.scope + formatName(refName)
	}

	return representation
}

// FormatKindScalarID formats an ID argument using the @expectedType directive.
// If expectedType is set, use the expected type name (as a pointer for objects,
// or as-is for interfaces). Otherwise fall back to the scalar name.
func (f *FormatTypeFunc) FormatKindScalarID(representation string, expectedType string) string {
	if expectedType == "" {
		// No expected type — use the raw ID type.
		representation += f.scope + formatName("ID")
		return representation
	}
	// Check if the expected type is an interface.
	if schema := generator.GetSchema(); schema != nil {
		if schemaType := schema.Types.Get(expectedType); schemaType != nil && schemaType.Kind == introspection.TypeKindInterface {
			representation += f.scope + formatName(expectedType)
			return representation
		}
	}
	representation += "*" + f.scope + formatName(expectedType)
	return representation
}

func (f *FormatTypeFunc) FormatKindObject(representation string, refName string, input bool) string {
	representation += f.scope + formatName(refName)
	return representation
}

func (f *FormatTypeFunc) FormatKindInputObject(representation string, refName string, input bool) string {
	representation += f.scope + formatName(refName)
	return representation
}

func (f *FormatTypeFunc) FormatKindEnum(representation string, refName string) string {
	representation += f.scope + refName
	return representation
}

File: sdk/rust/crates/dagger-codegen/src/rust/format.rs
use crate::functions::FormatTypeFuncs;

use super::functions::format_name;

pub struct FormatTypeFunc;

impl FormatTypeFuncs for FormatTypeFunc {
    fn format_kind_list(&self, representation: &str, _input: bool, _immutable: bool) -> String {
        format!("Vec<{}>", representation)
    }

    fn format_kind_scalar_string(&self, representation: &str, input: bool) -> String {
        let mut rep = representation.to_string();
        if input {
            rep.push_str("impl Into<String>");
        } else {
            rep.push_str("String");
        }
        rep
    }

    fn format_kind_scalar_int(&self, representation: &str) -> String {
        let mut rep = representation.to_string();
        rep.push_str("isize");
        rep
    }

    fn format_kind_scalar_float(&self, representation: &str) -> String {
        let mut rep = representation.to_string();
        rep.push_str("float");
        rep
    }

    fn format_kind_scalar_boolean(&self, representation: &str) -> String {
        let mut rep = representation.to_string();
        rep.push_str("bool");
        rep
    }

    fn format_kind_scalar_default(
        &self,
        representation: &str,
        ref_name: &str,
        _input: bool,
    ) -> String {
        let mut rep = representation.to_string();
        rep.push_str(&format_name(ref_name));
        rep
    }

    fn format_kind_object(&self, representation: &str, ref_name: &str) -> String {
        let mut rep = representation.to_string();
        rep.push_str(&format_name(ref_name));
        rep
    }

    fn format_kind_input_object(&self, representation: &str, ref_name: &str) -> String {
        let mut rep = representation.to_string();
        rep.push_str(&format_name(ref_name));
        rep
    }

    fn format_kind_enum(&self, representation: &str, ref_name: &str) -> String {
        let mut rep = representation.to_string();
        rep.push_str(&format_name(ref_name));
        rep
    }
}
